File: ts/src/control-plane/contract-probes/cli/extract.ts
/**
 * `autoctx probes extract` -- synthesize a contract-probe suite from a
 * harness-trace JSON file.
 *
 * A harness trace bundles both:
 *
 *   - `observations`: what actually happened during a recorded run
 *     (terminal exit code / stdout / stderr; the workdir's present files;
 *     observed service endpoints; emitted artifacts and their content).
 *   - `expectations`: what the operator declared SHOULD have happened
 *     (expected exit code; required / allowed / ignored files; required
 *     endpoints; per-artifact JSON-field / substring / line-ending
 *     expectations).
 *
 * The extractor joins them into a runnable `ContractProbeSuite`. The
 * slice-6 `autoctx probes check` runs the resulting suite.
 *
 * Slice 7 supports the four AC-728 slice-1 probe kinds (terminal,
 * directory, service, artifact). Cleanup / media / distributed kinds
 * land in follow-up slices once their trace formats are settled.
 *
 * Wire-format invariants mirror the suite runner (PR #990):
 *
 *   - Every nested object is `.strict()`; unknown keys fail validation.
 *   - RegExp values may be a bare string or `{ source, flags? }`; invalid
 *     regexes surface as Zod issues, not raw SyntaxError.
 *   - Date values transform from ISO-8601 strings; malformed dates
 *     surface as Zod issues.
 */

import { parseArgs } from "node:util";
import { readFileSync, writeFileSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";

import { z } from "zod";

import type { ContractProbeSuite } from "../index.js";

export interface ProbesExtractResult {
  readonly stdout: string;
  readonly stderr: string;
  readonly exitCode: number;
}

export const EXTRACT_HELP_TEXT = `autoctx probes extract -- synthesize a contract-probe suite from a harness trace.

Usage:
  autoctx probes extract --trace <path> [--output <path>]
  autoctx probes extract --help

A harness trace bundles observations (what happened in a recorded run)
and expectations (what the operator declared should have happened). The
extractor joins them into a runnable probe suite that \`autoctx probes
check\` can execute. See the "Contract Probes" section of the autoctx
README for the harness-trace JSON format and a minimal example.

Options:
  --trace <path>   Path to a harness-trace JSON file. Required.
  --output <path>  Write the resulting suite to this path. Parent
                   directories are created. If omitted, the suite is
                   emitted to stdout so it can be piped to
                   \`autoctx probes check --suite -\`.
  -h, --help       Show this help text.

Exit codes:
  0   the trace parsed and a suite was emitted.
  1   the trace failed to load / parse, or a write to --output failed.

The slice-7 extractor supports four probe kinds (terminal, directory,
service, artifact). Cleanup / media / distributed kinds are scaffolded
in follow-up slices once their trace formats settle.
`;

// --- Shared Zod helpers (mirroring runner.ts) ------------------------------

const RegExpJson = z
  .union([z.string(), z.object({ source: z.string(), flags: z.string().optional() }).strict()])
  .transform((value, ctx) => {
    try {
      return typeof value === "string" ? new RegExp(value) : new RegExp(value.source, value.flags);
    } catch (err) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `invalid regular expression: ${err instanceof Error ? err.message : String(err)}`,
      });
      return z.NEVER;
    }
  });

// --- Per-kind observation + expectation schemas ----------------------------

const TerminalObservation = z
  .object({
    exitCode: z.number(),
    stdout: z.string().default(""),
    stderr: z.string().default(""),
  })
  .strict();

const TerminalExpectations = z
  .object({
    expectedExitCode: z.number().optional(),
    requiredStdoutPatterns: z.array(RegExpJson).optional(),
    forbiddenStdoutPatterns: z.array(RegExpJson).optional(),
    requiredStderrPatterns: z.array(RegExpJson).optional(),
    forbiddenStderrPatterns: z.array(RegExpJson).optional(),
  })
  .strict();

const WorkdirObservation = z
  .object({
    presentFiles: z.array(z.string()),
  })
  .strict();

const DirectoryExpectations = z
  .object({
    requiredFiles: z.array(z.string()).default([]),
    allowedFiles: z.array(z.string()).default([]),
    ignoredPatterns: z.array(RegExpJson).optional(),
  })
  .strict();

const ServiceEndpoint = z
  .object({
    host: z.string(),
    port: z.number(),
    protocol: z.enum(["tcp", "udp"]).optional(),
  })
  .strict();

const ServiceExpectations = z
  .object({
    required: z.array(ServiceEndpoint).default([]),
    allowed: z.array(ServiceEndpoint).optional(),
  })
  .strict();

const ArtifactObservation = z
  .object({
    path: z.string(),
    content: z.string(),
  })
  .strict();

const ArtifactExpectations = z
  .object({
    path: z.string(),
    label: z.string().optional(),
    expectedLineEnding: z.enum(["lf", "crlf"]).optional(),
    requiredSubstrings: z.array(z.string()).optional(),
    forbiddenSubstrings: z.array(z.string()).optional(),
    requiredJsonFields: z.array(z.string()).optional(),
  })
  .strict();

// --- HarnessTrace envelope -------------------------------------------------

const HarnessObservationsSchema = z
  .object({
    terminal: TerminalObservation.optional(),
    workdir: WorkdirObservation.optional(),
    services: z.array(ServiceEndpoint).optional(),
    artifacts: z.array(ArtifactObservation).optional(),
  })
  .strict();

const HarnessExpectationsSchema = z
  .object({
    terminal: TerminalExpectations.optional(),
    directory: DirectoryExpectations.optional(),
    services: ServiceExpectations.optional(),
    artifacts: z.array(ArtifactExpectations).optional(),
  })
  .strict();

export const HarnessTraceSchema = z
  .object({
    schema_version: z.literal(1),
    label: z.string().optional(),
    observations: HarnessObservationsSchema,
    expectations: HarnessExpectationsSchema.optional(),
  })
  .strict()
  // PR #992 review (P2): every declared expectation must have its matching
  // observation. Without this guard, an expectation-only section was
  // silently dropped at extraction time and the resulting suite passed
  // vacuously -- the same silent-pass class of bug the slice-5
  // `missing-observation` failure kind was added to close.
  .superRefine((trace, ctx) => {
    const obs = trace.observations;
    const exp = trace.expectations;
    if (exp === undefined) {
      return;
    }
    if (exp.terminal !== undefined && obs.terminal === undefined) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["expectations", "terminal"],
        message:
          "expectation declared without a matching observation; add `observations.terminal` to record exit code / stdout / stderr",
      });
    }
    if (exp.directory !== undefined && obs.workdir === undefined) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["expectations", "directory"],
        message:
          "expectation declared without a matching observation; add `observations.workdir` to record present files",
      });
    }
    if (exp.services !== undefined && obs.services === undefined) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["expectations", "services"],
        message:
          "expectation declared without a matching observation; add `observations.services` to record observed endpoints",
      });
    }
    if (exp.artifacts !== undefined) {
      if (obs.artifacts === undefined) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          path: ["expectations", "artifacts"],
          message:
            "per-artifact expectations declared without `observations.artifacts`; add the matching artifact observations",
        });
      } else {
        // Per-artifact: every expectation must reference a path that the
        // observations actually contain.
        const observedPaths = new Set(obs.artifacts.map((a) => a.path));
        exp.artifacts.forEach((artExp, index) => {
          if (!observedPaths.has(artExp.path)) {
            ctx.addIssue({
              code: z.ZodIssueCode.custom,
              path: ["expectations", "artifacts", index, "path"],
              message: `expectation references artifact path ${JSON.stringify(
                artExp.path,
              )} but no observation with that path was recorded`,
            });
          }
        });
      }
    }
  });

export type HarnessTrace = z.infer<typeof HarnessTraceSchema>;

// --- Extractor -------------------------------------------------------------

/** Build a `ContractProbeSuite` (slice 5 wire shape) from a harness trace.
 * Pure function: no IO. The CLI handler wraps this with file IO + arg
 * parsing. */
export function extractContractProbeSuite(trace: HarnessTrace): ContractProbeSuite {
  const probes: ContractProbeSuite["probes"] = [];
  const expectations = trace.expectations ?? {};
  const label = trace.label;

  if (trace.observations.terminal !== undefined) {
    const exp = expectations.terminal ?? {};
    probes.push({
      kind: "terminal",
      ...(label !== undefined ? { label } : {}),
      inputs: {
        exitCode: trace.observations.terminal.exitCode,
        stdout: trace.observations.terminal.stdout,
        stderr: trace.observations.terminal.stderr,
        ...(exp.expectedExitCode !== undefined ? { expectedExitCode: exp.expectedExitCode } : {}),
        ...(exp.requiredStdoutPatterns !== undefined
          ? { requiredStdoutPatterns: exp.requiredStdoutPatterns }
          : {}),
        ...(exp.forbiddenStdoutPatterns !== undefined
          ? { forbiddenStdoutPatterns: exp.forbiddenStdoutPatterns }
          : {}),
        ...(exp.requiredStderrPatterns !== undefined
          ? { requiredStderrPatterns: exp.requiredStderrPatterns }
          : {}),
        ...(exp.forbiddenStderrPatterns !== undefined
          ? { forbiddenStderrPatterns: exp.forbiddenStderrPatterns }
          : {}),
      },
    });
  }

  if (trace.observations.workdir !== undefined) {
    const exp = expectations.directory ?? { requiredFiles: [], allowedFiles: [] };
    probes.push({
      kind: "directory",
      ...(label !== undefined ? { label } : {}),
      inputs: {
        presentFiles: trace.observations.workdir.presentFiles,
        requiredFiles: exp.requiredFiles,
        allowedFiles: exp.allowedFiles,
        ...(exp.ignoredPatterns !== undefined ? { ignoredPatterns: exp.ignoredPatterns } : {}),
      },
    });
  }

  if (trace.observations.services !== undefined) {
    const exp = expectations.services ?? { required: [] };
    probes.push({
      kind: "service",
      ...(label !== undefined ? { label } : {}),
      inputs: {
        observed: trace.observations.services,
        required: exp.required,
        ...(exp.allowed !== undefined ? { allowed: exp.allowed } : {}),
      },
    });
  }

  if (trace.observations.artifacts !== undefined) {
    // Per-artifact: match each observation by `path` against an expectations
    // entry; absent expectations leave the artifact probe in a no-op shape
    // (path + content only, no required/forbidden lists).
    const expectationsByPath = new Map<string, z.infer<typeof ArtifactExpectations>>();
    for (const artExp of expectations.artifacts ?? []) {
      expectationsByPath.set(artExp.path, artExp);
    }
    for (const artifact of trace.observations.artifacts) {
      const exp = expectationsByPath.get(artifact.path);
      probes.push({
        kind: "artifact",
        ...(exp?.label !== undefined ? { label: exp.label } : label !== undefined ? { label } : {}),
        inputs: {
          path: artifact.path,
          content: artifact.content,
          ...(exp?.expectedLineEnding !== undefined
            ? { expectedLineEnding: exp.expectedLineEnding }
            : {}),
          ...(exp?.requiredSubstrings !== undefined
            ? { requiredSubstrings: exp.requiredSubstrings }
            : {}),
          ...(exp?.forbiddenSubstrings !== undefined
            ? { forbiddenSubstrings: exp.forbiddenSubstrings }
            : {}),
          ...(exp?.requiredJsonFields !== undefined
            ? { requiredJsonFields: exp.requiredJsonFields }
            : {}),
        },
      });
    }
  }

  return { schema_version: 1, probes };
}

/** Serialize a `ContractProbeSuite` back to JSON suitable for the suite
 * runner. RegExp values are emitted as `{ source, flags }` objects (the
 * runner schema accepts both bare strings and that object form). */
function serializeSuite(suite: ContractProbeSuite): string {
  return JSON.stringify(
    suite,
    (_key, value) => {
      if (value instanceof RegExp) {
        return { source: value.source, flags: value.flags };
      }
      return value;
    },
    2,
  );
}

// --- CLI handler -----------------------------------------------------------

interface ParsedArgs {
  readonly tracePath?: string;
  readonly outputPath?: string;
  readonly help: boolean;
}

function parseExtractArgs(args: readonly string[]): ParsedArgs | { readonly error: string } {
  try {
    const { values } = parseArgs({
      args: [...args],
      options: {
        trace: { type: "string" },
        output: { type: "string" },
        help: { type: "boolean", short: "h", default: false },
      },
      allowPositionals: false,
    });
    return {
      tracePath: typeof values.trace === "string" ? values.trace : undefined,
      outputPath: typeof values.output === "string" ? values.output : undefined,
      help: values.help === true,
    };
  } catch (err) {
    return { error: err instanceof Error ? err.message : String(err) };
  }
}

export async function runExtract(args: readonly string[]): Promise<ProbesExtractResult> {
  const parsed = parseExtractArgs(args);
  if ("error" in parsed) {
    return {
      stdout: "",
      stderr: `autoctx probes extract: ${parsed.error}\n\n${EXTRACT_HELP_TEXT}`,
      exitCode: 1,
    };
  }

  if (parsed.help) {
    return { stdout: EXTRACT_HELP_TEXT, stderr: "", exitCode: 0 };
  }

  if (parsed.tracePath === undefined || parsed.tracePath.length === 0) {
    return {
      stdout: "",
      stderr: `autoctx probes extract: --trace <path> is required\n\n${EXTRACT_HELP_TEXT}`,
      exitCode: 1,
    };
  }

  let raw: string;
  try {
    raw = readFileSync(parsed.tracePath, "utf-8");
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return {
      stdout: "",
      stderr: `autoctx probes extract: failed to read trace from ${parsed.tracePath}: ${message}`,
      exitCode: 1,
    };
  }

  let raw_parsed: unknown;
  try {
    raw_parsed = JSON.parse(raw);
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return {
      stdout: "",
      stderr: `autoctx probes extract: invalid JSON in ${parsed.tracePath}: ${message}`,
      exitCode: 1,
    };
  }

  const validation = HarnessTraceSchema.safeParse(raw_parsed);
  if (!validation.success) {
    const issues = validation.error.issues
      .map((issue) => {
        const path = issue.path.length > 0 ? issue.path.join(".") : "<root>";
        return `  - ${path}: ${issue.message}`;
      })
      .join("\n");
    return {
      stdout: "",
      stderr: `autoctx probes extract: trace validation failed\n${issues}`,
      exitCode: 1,
    };
  }

  const suite = extractContractProbeSuite(validation.data);
  const serialized = serializeSuite(suite);

  if (parsed.outputPath !== undefined && parsed.outputPath.length > 0) {
    try {
      mkdirSync(dirname(parsed.outputPath), { recursive: true });
      writeFileSync(parsed.outputPath, serialized + "\n", "utf-8");
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      return {
        stdout: "",
        stderr: `autoctx probes extract: failed to write suite to ${parsed.outputPath}: ${message}`,
        exitCode: 1,
      };
    }
    return { stdout: `wrote suite to ${parsed.outputPath}`, stderr: "", exitCode: 0 };
  }

  return { stdout: serialized, stderr: "", exitCode: 0 };
}

File: autocontext/src/autocontext/control_plane/contract_probes/extract.py
"""AC-728 `autoctx probes extract` library surface (Python parity, slice 5).

Mirrors the slice-1 / TS PR #992 portion of
``ts/src/control-plane/contract-probes/cli/extract.ts``: reads a
harness-trace JSON envelope and synthesises a runnable
``ContractProbeSuite`` covering the four slice-1 probe kinds
(terminal / directory / service / artifact). The advanced kinds
(cleanup / media / distributed) land in slice 6.

A harness trace bundles two halves:

- ``observations``: what actually happened in a recorded run.
- ``expectations``: what the operator declared should have happened.

The extractor joins them. Per the slice-1 audit invariant, every
declared expectation must have a matching observation; orphan
expectations fail validation at parse time rather than silently
producing a vacuously-passing suite.

Module split lives next to ``runner.py`` so both load the same
``ContractProbeSuite`` and ``_PatternList`` helpers.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Literal

from pydantic import (
    StrictInt,
    model_validator,
)

from .runner import (
    _PatternList,
    _Strict,
)

__all__ = [
    "HarnessTrace",
    "HarnessTraceSchema",
    "extract_contract_probe_suite",
    "load_harness_trace",
    "serialize_suite",
]


# ---------------------------------------------------------------------------
# Per-kind observation + expectation schemas
# ---------------------------------------------------------------------------


class _TerminalObservation(_Strict):
    exitCode: StrictInt
    stdout: str = ""
    stderr: str = ""


class _TerminalExpectations(_Strict):
    expectedExitCode: StrictInt | None = None
    requiredStdoutPatterns: _PatternList | None = None
    forbiddenStdoutPatterns: _PatternList | None = None
    requiredStderrPatterns: _PatternList | None = None
    forbiddenStderrPatterns: _PatternList | None = None


class _WorkdirObservation(_Strict):
    presentFiles: tuple[str, ...]


class _DirectoryExpectations(_Strict):
    requiredFiles: tuple[str, ...] = ()
    allowedFiles: tuple[str, ...] = ()
    ignoredPatterns: _PatternList | None = None


class _ServiceEndpointSchema(_Strict):
    host: str
    port: StrictInt
    protocol: Literal["tcp", "udp"] | None = None


class _ServiceExpectations(_Strict):
    required: tuple[_ServiceEndpointSchema, ...] = ()
    allowed: tuple[_ServiceEndpointSchema, ...] | None = None


class _ArtifactObservation(_Strict):
    path: str
    content: str


class _ArtifactExpectations(_Strict):
    path: str
    label: str | None = None
    expectedLineEnding: Literal["lf", "crlf"] | None = None
    requiredSubstrings: tuple[str, ...] | None = None
    forbiddenSubstrings: tuple[str, ...] | None = None
    requiredJsonFields: tuple[str, ...] | None = None


# ---------------------------------------------------------------------------
# Trace envelope
# ---------------------------------------------------------------------------


class _HarnessObservations(_Strict):
    terminal: _TerminalObservation | None = None
    workdir: _WorkdirObservation | None = None
    services: tuple[_ServiceEndpointSchema, ...] | None = None
    artifacts: tuple[_ArtifactObservation, ...] | None = None


class _HarnessExpectations(_Strict):
    terminal: _TerminalExpectations | None = None
    directory: _DirectoryExpectations | None = None
    services: _ServiceExpectations | None = None
    artifacts: tuple[_ArtifactExpectations, ...] | None = None


class HarnessTrace(_Strict):
    """Slice-5 harness-trace envelope: 4 base probe kinds only.

    Slice 6 will extend ``observations`` / ``expectations`` with
    cleanup / media / distributed sections. The envelope itself
    forbids unknown keys today, so adding sections in slice 6 is a
    visible source change.
    """

    schema_version: Literal[1]
    label: str | None = None
    observations: _HarnessObservations
    expectations: _HarnessExpectations | None = None

    @model_validator(mode="after")
    def _check_orphan_expectations(self) -> HarnessTrace:
        """Every declared expectation must have its matching observation.

        Mirrors the TS ``superRefine`` (PR #992 review P2): without
        this guard an expectation-only section would be silently
        dropped at extraction time and the resulting suite would
        pass vacuously. The slice-1 audit invariant says: corrupted
        traces fail loudly, not silently.
        """
        if self.expectations is None:
            return self

        violations: list[str] = []
        obs = self.observations
        exp = self.expectations

        if exp.terminal is not None and obs.terminal is None:
            violations.append(
                "expectations.terminal: declared without observations.terminal; add terminal exit code / stdout / stderr"
            )
        if exp.directory is not None and obs.workdir is None:
            violations.append("expectations.directory: declared without observations.workdir; add the present-files list")
        if exp.services is not None and obs.services is None:
            violations.append("expectations.services: declared without observations.services; add the observed endpoints list")

        if exp.artifacts is not None:
            if obs.artifacts is None:
                violations.append(
                    "expectations.artifacts: per-artifact expectations declared "
                    "without observations.artifacts; add the matching observations"
                )
            else:
                observed_paths = {a.path for a in obs.artifacts}
                seen_paths: set[str] = set()
                for index, art_exp in enumerate(exp.artifacts):
                    if art_exp.path not in observed_paths:
                        violations.append(
                            f"expectations.artifacts.{index}.path: expectation "
                            f"references {art_exp.path!r} but no observation "
                            "with that path was recorded"
                        )
                    if art_exp.path in seen_paths:
                        violations.append(
                            f"expectations.artifacts.{index}.path: duplicate "
                            f"per-artifact expectation for {art_exp.path!r}; "
                            "merge into a single entry"
                        )
                    seen_paths.add(art_exp.path)

        if violations:
            raise ValueError("\n".join(violations))
        return self


HarnessTraceSchema = HarnessTrace


# ---------------------------------------------------------------------------
# Extractor
# ---------------------------------------------------------------------------


def _service_endpoint_dict(endpoint: _ServiceEndpointSchema) -> dict:
    payload = {"host": endpoint.host, "port": endpoint.port}
    if endpoint.protocol is not None:
        payload["protocol"] = endpoint.protocol
    return payload


def _patterns_dict_list(patterns: tuple) -> list[dict]:
    """Serialise a ``_PatternList`` as a list of ``{source, flags}`` dicts.

    The runner schema accepts either bare strings or ``{source, flags}``
    objects. Emitting the dict form preserves any flags set on the
    compiled pattern.
    """
    out: list[dict] = []
    flag_chars = {
        2: "i",  # re.IGNORECASE
        8: "m",  # re.MULTILINE
        16: "s",  # re.DOTALL
        64: "x",  # re.VERBOSE
        32: "u",  # re.UNICODE
    }
    for pat in patterns:
        flags = ""
        for bit, ch in flag_chars.items():
            if pat.flags & bit:
                flags += ch
        # re.UNICODE is set by default in Python 3 for str patterns; drop
        # it from the serialised flag string so round-trips do not gain
        # a spurious `u`.
        flags = flags.replace("u", "")
        entry: dict = {"source": pat.pattern}
        if flags:
            entry["flags"] = flags
        out.append(entry)
    return out


def extract_contract_probe_suite(trace: HarnessTrace) -> dict:
    """Build a ``ContractProbeSuite`` dict from a harness trace.

    Returns a plain dict matching the slice-3 suite wire format so
    callers can serialise it directly via ``json.dumps`` without an
    intermediate model dump. Pure function; no IO.
    """
    probes: list[dict] = []
    label = trace.label
    expectations = trace.expectations

    if trace.observations.terminal is not None:
        term_exp = expectations.terminal if expectations else None
        inputs: dict = {
            "exitCode": trace.observations.terminal.exitCode,
            "stdout": trace.observations.terminal.stdout,
            "stderr": trace.observations.terminal.stderr,
        }
        if term_exp is not None:
            if term_exp.expectedExitCode is not None:
                inputs["expectedExitCode"] = term_exp.expectedExitCode
            if term_exp.requiredStdoutPatterns is not None:
                inputs["requiredStdoutPatterns"] = _patterns_dict_list(term_exp.requiredStdoutPatterns)
            if term_exp.forbiddenStdoutPatterns is not None:
                inputs["forbiddenStdoutPatterns"] = _patterns_dict_list(term_exp.forbiddenStdoutPatterns)
            if term_exp.requiredStderrPatterns is not None:
                inputs["requiredStderrPatterns"] = _patterns_dict_list(term_exp.requiredStderrPatterns)
            if term_exp.forbiddenStderrPatterns is not None:
                inputs["forbiddenStderrPatterns"] = _patterns_dict_list(term_exp.forbiddenStderrPatterns)
        probe: dict = {"kind": "terminal", "inputs": inputs}
        if label is not None:
            probe["label"] = label
        probes.append(probe)

    if trace.observations.workdir is not None:
        dir_exp = expectations.directory if expectations else None
        inputs = {
            "presentFiles": list(trace.observations.workdir.presentFiles),
            "requiredFiles": list(dir_exp.requiredFiles) if dir_exp else [],
            "allowedFiles": list(dir_exp.allowedFiles) if dir_exp else [],
        }
        if dir_exp is not None and dir_exp.ignoredPatterns is not None:
            inputs["ignoredPatterns"] = _patterns_dict_list(dir_exp.ignoredPatterns)
        probe = {"kind": "directory", "inputs": inputs}
        if label is not None:
            probe["label"] = label
        probes.append(probe)

    if trace.observations.services is not None:
        svc_exp = expectations.services if expectations else None
        inputs = {
            "observed": [_service_endpoint_dict(e) for e in trace.observations.services],
            "required": [_service_endpoint_dict(e) for e in svc_exp.required] if svc_exp is not None else [],
        }
        if svc_exp is not None and svc_exp.allowed is not None:
            inputs["allowed"] = [_service_endpoint_dict(e) for e in svc_exp.allowed]
        probe = {"kind": "service", "inputs": inputs}
        if label is not None:
            probe["label"] = label
        probes.append(probe)

    if trace.observations.artifacts is not None:
        artifact_exp_by_path: dict[str, _ArtifactExpectations] = {}
        if expectations is not None and expectations.artifacts is not None:
            for declared_exp in expectations.artifacts:
                artifact_exp_by_path[declared_exp.path] = declared_exp
        for artifact in trace.observations.artifacts:
            art_exp = artifact_exp_by_path.get(artifact.path)
            inputs = {"path": artifact.path, "content": artifact.content}
            if art_exp is not None:
                if art_exp.expectedLineEnding is not None:
                    inputs["expectedLineEnding"] = art_exp.expectedLineEnding
                if art_exp.requiredSubstrings is not None:
                    inputs["requiredSubstrings"] = list(art_exp.requiredSubstrings)
                if art_exp.forbiddenSubstrings is not None:
                    inputs["forbiddenSubstrings"] = list(art_exp.forbiddenSubstrings)
                if art_exp.requiredJsonFields is not None:
                    inputs["requiredJsonFields"] = list(art_exp.requiredJsonFields)
            probe = {"kind": "artifact", "inputs": inputs}
            probe_label = art_exp.label if art_exp is not None and art_exp.label is not None else label
            if probe_label is not None:
                probe["label"] = probe_label
            probes.append(probe)

    return {"schema_version": 1, "probes": probes}


def serialize_suite(suite: dict) -> str:
    """Serialise an extracted suite to JSON suitable for the suite runner."""
    return json.dumps(suite, indent=2)


def load_harness_trace(path: str | Path) -> HarnessTrace:
    """Load a harness trace JSON file and validate it."""
    raw = Path(path).read_text(encoding="utf-8")
    parsed = json.loads(raw)
    return HarnessTraceSchema.model_validate(parsed)
